### ISSUE TITLE
[BUG][TGL-012-drop-stable] DSM Multicore Stream Stall

### ISSUE THREAD

**Issue by zrombel** [thesofproject/sof#3947]
**Describe the bug**
In DSM test where DSM component is run on CORE 1 and rest of topology on default CORE 0 after triggering pipelines start, playback pipeline stops copying data.

**Topology**
```
    +----------------------------------------------+
    | +------+   +---+   +- -+   +---+   +-------+ |
    | | Host |-->|Buf|-->|Dsm|-->|Buf|-->|SSP Dai|-------+
    | +------+   +---+   +---+   +---+   +-------+ |     |
    +------------------+   ^   +-------------------+     |
                       |   |   |                         |
                       | +---+ |                         |
                       | |Buf| |                         |
                       | +---+ |                         |
                       +---^---+                         |
    +----------------------|------------------------+    |
    | +------+   +---+   +-|--+   +---+   +-------+ |    |
    | | Host |<--|Buf|<--|Dmux|<--|Buf|<--|SSP Dai|<-----+
    | +------+   +---+   +----+   +---+   +-------+ |
    +-----------------------------------------------+
```

**To Reproduce**
Run Python test:
22_05_TestDsmMulticoreSsp48000Hz24b32b2ch or 22_05_TestDsmMulticoreSsp48000Hz32b32b2ch
with parameters:
--playback-iterations=100

**Reproduction Rate**
~1/10

**Environment**
1) Name of the platform(s) on which the bug is observed.
    * Platform: TGL B0 RVP
2) Firmware branch name and commit 
    * Branch: tgl-012-drop-stable
    * Hash: fa35ee0
3) Python tests branch name and commit
    * Branch: master
    * Hash: 237e3511f3787aa9b45c53b43bea33cfa2646114

**Logs**
[22_05_TestDsmMulticoreSsp48000Hz32b32b2ch FAIL.zip](https://github.com/thesofproject/sof/files/6188732/22_05_TestDsmMulticoreSsp48000Hz32b32b2ch.FAIL.zip)

**Comment by slawblauciak** [comment]
I was able to bisect and find which change introduces this behavior:
https://github.com/thesofproject/sof/commit/5104f5e02df5a720f2896d8403a880cde286f657
However, the same change exists on tgl-011-drop-stable branch, and the problem does not occur there.
This leads me to believe that this behavior is a result of at least two changes. 
I suspect the problem might have its source in the scheduler again, as such I will continue debugging from that side.

**Comment by keyonjie** [comment]
Hi @slawblauciak @zrombel  can you try if reverting https://github.com/thesofproject/sof/commit/1fa1001e24a8bc8db419e8ff97aaf591ae8ceee5 will help?

**Comment by lgirdwood** [comment]
@keyonjie @slawblauciak any updates here ?

**Comment by slawblauciak** [comment]
@lgirdwood I was only able to eliminate some possibilities: it's most likely not related to IDC or spin locks. I had some trace points set up around those and there were no oddities there. Additionally, when the issue occurs I can see that none of the pipelines actually work, the pipeline tasks are not called.
I'll be debugging it further from scheduler side.

**Comment by slawblauciak** [comment]
Also, neglected to mention that this issue is only reproduced with Python tests under slim driver. No attempt of reproduction under full stack was successful.

**Comment by lgirdwood** [comment]
@slawblauciak thanks, btw how is core 1 work scheduled and synchronized with core 0 in this case ?

**Comment by slawblauciak** [comment]
@lgirdwood good question, I'm not exactly familiar with the way SMP is implemented in the firmware.

**Comment by lgirdwood** [comment]
@slawblauciak the issue I'm thinking here if core 1 work takes too long for one iteration and how this could spill over into the core0 work (and how we could detect/report and recover)..

**Comment by slawblauciak** [comment]
@lgirdwood that was my initial though as well, but I believe the work on the cores is scheduled entirely independently and without synchronization, except for the locks in the buffer access. But again, I'm not 100% sure.

**Comment by slawblauciak** [comment]
Upon further debug I was able to partly determine source of the problem. There was a small discrepancy between latest changes made to the scheduler, between tgl-011 and tgl-012. Upon aligning the changes, I was able to get the scheduler to run somewhat properly now. However, the issue still occurs, which suggests that the task might be being incorrectly cancelled or erroneously handled otherwise. Unfortunately I lack the required understanding of the scheduler as a whole and the changes made to be able to tell why aligning them improves the behavior at this moment.
